addon/components/au-file-upload.js

```diff
@@ -67,43 +69,57 @@ export default class AuFileUploadComponent extends Component {
       const response = yield file.upload(this.endPoint, {
         'Content-Type': 'multipart/form-data',
       });
-      const fileId = response.body?.data?.id;
+      const body = yield response.json();
+      const fileId = body?.data?.id;
       return fileId;
     } catch (e) {
-      this.uploadErrorData = [...this.uploadErrorData, { filename: file.name }];
+      this.addError(file);
       this.removeFileFromQueue(file);
       return null;
     }
   }
 
-  hasValidationErrors(file) {
-    // TODO: see if we can split the side-effects from the
-    // non-side-effects.  The name doesn't suggest this will yield a
-    // state change.
-    if (file.size > this.maxFileSizeMB * Math.pow(1024, 2)) {
-      this.uploadErrorData = [
-        ...this.uploadErrorData,
-        {
-          filename: file.name,
-          error: `Bestand is te groot (max ${this.maxFileSizeMB} MB)`,
-        },
-      ];
-      this.removeFileFromQueue(file);
-      return true;
+  @action
+  filter(file, files, index) {
+    let isFirstFile = index === 0;
+
+    if (isFirstFile) {
+      this.resetErrors();
+    } else {
+      if (!this.args.multiple) {
+        // We only upload the first file if `@multiple` is not set to true. This matches the behavior of ember-file-upload v4.
+        return false;
+      }
     }
-    return false;
-  }
 
-  @action
-  resetErrors() {
-    this.uploadErrorData = [];
+    if (!isValidFileSize(file.size, this.maxFileSizeMB)) {
+      this.addError(file, `Bestand is te groot (max ${this.maxFileSizeMB} MB)`);
+      return false;
+    }
+
+    if (!isValidFileType(file, this.args.accept)) {
+      this.addError(file, this.helpTextFileNotSupported);
+      return false;
+    }
+
+    if (typeof this.args.filter === 'function') {
+      let shouldBeUploaded = this.args.filter(...arguments);
+
+      if (shouldBeUploaded) {
+        return true;
+      } else {
+        this.addError(file, this.helpTextFileNotSupported);
+        return false;
+      }
+    } else {
+      return true;
+    }
```

Comment:
At the moment it is possible to add extra validations on top by passing the `@filter` argument. The current implementation doesn't let you return a validation message though. It sounds nice to allow that? 

Some options:
- let `@filter` return a boolean _or_ a string. `true` = valid, `false` = invalid (use current logic), `"string"` = invalid, use custom message
- let `@filter` return an object `{ isValid: boolean, error?: string }`

Any preferences or other suggestions?

Reply:
It might also be better to rename `@filter` to something else since it doesn't fully match the `ember-file-upload` version anymore.

Reply:
Or we drop the `@filter` functionality for now since it's a now feature and this was just meant to be an update. I initially didn't implement the v4 validations so the filter was the alternative, but now that's not needed anymore. I'll go that route for now and open a feature request issue for the custom validations.